### ISSUE TITLE
Reproduce : BLEU is 10+ lower  than reported

### ISSUE THREAD

**Issue by EricLina** [ictnlp/GMA#2]
# GMA WMT15 German-to-English with Transformer-Base : delta=2.5 
@Vily1998 
I've followed the guidance and trained the model for about 100 epoch using following set:
```
delta=2.5 
python train.py --ddp-backend=no_c10d ${data} --arch transformer --share-all-embeddings \
 --optimizer adam \
 --adam-betas '(0.9, 0.98)' \
 --clip-norm 0.0 \
 --lr 5e-4 \
 --lr-scheduler inverse_sqrt \
 --warmup-init-lr 1e-07 \
 --warmup-updates 4000 \
 --dropout 0.3 \
 --criterion label_smoothed_cross_entropy \
 --label-smoothing 0.1 \
 --left-pad-source False \
 --delta ${delta} \
 --save-dir ${modelfile} \
 --max-tokens 4096 --update-freq 2
``` 

than I evaluate it on the perl script you provided by single checkpoint and average the last 4 cp .

Below is my result :
| CheckPointNum            | BLEU  |
|--------------------------|-------|
| 5                        | 17.52 |
| 10                       | 18.09 |
| 20                       | 18.83 |
| 25                       | 18.66 |
| 40                       | 18.93 |
| 50                       | 18.80 |
| 60                       | 19.03 |
| 70                       | 18.99 |
| 80                       | 18.72 |
| 90                       | 18.83 |
| Average for 85,90,95,100 | 18.47 |

That is lower than reported (BLEU: 31.91) ,But I don't know the reason. So I am here to check something.

# Questions
1.  **what should the translation reference look like** ?
     when I generate for BLEU , it would get an error :
> It is not advisable to publish scores from multi-bleu.perl.  The scores depend on your tokenizer, which is unlikely to be reproducible from your paper or consistent across research groups.  Instead you should detokenize then use mteval-v14.pl, which has a standard tokenization.  Scores from multi-bleu.perl can still be used for internal purposes when you have a consistent tokenizer.

My translation and reference is just like below :
```
> head -5 $ref_dir
India and Japan prime ministers meet in Tokyo
India &apos;s new prime minister , Nar@@ end@@ ra Mo@@ di , is meeting his Japanese counter@@ part , Shin@@ zo A@@ be , in Tokyo to discuss economic and security ties , on his first major foreign visit since winning May &apos;s election .
Mr Mo@@ di is on a five @-@ day trip to Japan to strengthen economic ties with the third largest economy in the world .
High on the agenda are plans for greater nuclear co @-@ operation .
India is also re@@ porte@@ dly hoping for a deal on defence collaboration between the two nations .

>head -5 $pred.translation
The prime ministers of India and Japan met in Tokyo .
India &apos;s new prime minister , Narendra Modi , meets in his first major foreign visit since his victory in May , his Japanese counterparts Shinzo Abe in Tokyo to discuss economic and security policy .
Mr Modi is on a five @-@ day trip to Japan to improve economic relations with the third largest economic nation in the world .
Plans for greater nuclear cooperation are high on the agenda .
Reports suggest that India also hopes for a treaty on defence cooperation between the two nations .
```
I'am not sure that whether it is right ?

2. **How mant epoch it need to train ?**
As you can see , I've trained it for about 100 epoches and the BLUE seems has converged .In your experiments at the time, how long /how many epoches did you train with this set of hyperparameters? Using the provided  set of hyperparameters , Is it possible to ensure that the effect is reproduced?

3. **Does torch1.11 support  GMA ?**
I noticed many torch operation is for torch old version , and there are many warning like below , I'am sorry that my torch version doesn't match ,I will try an old version later
>UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  beams_buf = indices_buf // vocab_size

>/GMA/fairseq/optim/adam.py:173: UserWarning: This overload of add_ is deprecated:
        add_(Number alpha, Tensor other)  
Consider using one of the following signatures instead:
        add_(Tensor other, *, Number alpha) (Triggered internally at  /opt/conda/conda-bld/pytorch_1603729138878/work/torch/csrc/utils/python_arg_parser.cpp:882.)
  exp_avg.mul_(beta1).add_(1 - beta1, grad)

Ps: I've changed my torch version to the right one and ran it again , the result seemed same as before , so the torch version 
may not affect the result. 

### Env info
| NAME | Version |
| --- | ----------- |
| python | 3.7 |
| torch| 1.7.0+cu10.2|
|fairseq|0.8.0|

**Comment by zhangshaolei1998** [comment]
1. The output warning is ignorable.
2. 300k updates(60-70 epoches) is enough for the model to converge. The results can be reproduced with the recommended hyperparameters. Especially we suggest set delta=1.0, but your script set delta=2.5. 
With a difference of 10+BLEU, we strongly recommend that you train a full-sentence MT model first to ensure that there are no errors in the training process or the use of fairseq. Make sure full-sentence MT has a BLEU score greater than 30.
3. Later versions of Pytorch may have warnings due to some deprecations, but it does not affect training of GMA.

**Comment by EricLina** [comment]
Thank you. I've nearly solve it.
The reason why the bleu is lower 10+ is that, When evaluating ,I did not using the detokenized and removed bpe for reference.


**here is what I used before**
>head -5 pred.translation
The prime ministers of India and Japan met in Tokyo .
India &apos;s new prime minister , Narendra Modi , meets in his first major foreign visit since his victory in May , his Japanese counterparts Shinzo Abe in Tokyo to discuss economic and security policy .
Mr Modi is on a five @-@ day trip to Japan to improve economic relations with the third largest economic nation in the world .
Plans for greater nuclear cooperation are high on the agenda .
Reports suggest that India also hopes for a treaty on defence cooperation between the two nations .

>  head -5 $ref_dir
India and Japan prime ministers meet in Tokyo
India &apos;s new prime minister , Nar@@ end@@ ra Mo@@ di , is meeting his Japanese counter@@ part , Shin@@ zo A@@ be , in Tokyo to discuss economic and security ties , on his first major foreign visit since winning May &apos;s election .
Mr Mo@@ di is on a five @-@ day trip to Japan to strengthen economic ties with the third largest economy in the world .
High on the agenda are plans for greater nuclear co @-@ operation .
India is also re@@ porte@@ dly hoping for a deal on defence collaboration between the two nations .

**Actually I should remove bpe for the $ref_dir** 
>head -5 pred.translation
The prime ministers of India and Japan met in Tokyo .
India &apos;s new prime minister , Narendra Modi , meets with the US government &apos;s Japanese counterpart in his first major foreign visit since his May election victory , in May to discuss economic and security policy relations .
Mr Modi is on a five @-@ day trip to Japan to improve economic relations with the third largest economic nation in the world .
Plans for greater nuclear cooperation are high on the agenda .
According to reports , India also hopes for a treaty on defence cooperation between the two nations .

>head -5  $ref_dir
India and Japan prime ministers meet in Tokyo
India &apos;s new prime minister , Narendra Modi , is meeting his Japanese counterpart , Shinzo Abe , in Tokyo to discuss economic and security ties , on his first major foreign visit since winning May &apos;s election .
Mr Modi is on a five @-@ day trip to Japan to strengthen economic ties with the third largest economy in the world .
High on the agenda are plans for greater nuclear co @-@ operation .
India is also reportedly hoping for a deal on defence collaboration between the two nations .

I suggest that we should  complete the details of the data preparation and evaluation part .

All in all , The bleu result is reaching 29 now and it shows your code works ! Thank you !
@Vily1998
